alpa/pipeline_parallel/stage_profiling.py

```diff
@@ -493,115 +570,480 @@ def profile_all(stages, compiled_outputs: Sequence[CompileOutput], meshes,
+                    stage_indices = (start, end, mesh_id, config_idx)
+                    stages.append(
+                        (stage_indices, stage_config, autosharding_config))
+    return stages
+
+
+def get_max_n_succ_stages(profile_results: Sequence[ProfileResult]):
+    env = {}
+    initial_size = sum(result.initial_size for result in profile_results)
+    peak_memory = 0
+    available_memory = profile_results[0].available_memory
+    assert all(result.available_memory == available_memory
+               for result in profile_results)
+    n_modules = profile_results[0].n_modules
+    assert n_modules == 2, "Only support forward and backward module"
+    assert all(result.n_modules == n_modules for result in profile_results)
+
+    stage_no = n_modules * len(profile_results)
+    # eliminate_time[var] = k means that the variable can be eliminated after
+    # stage k.
+    eliminate_time = {}
+    required_outvars = set()
+    for i in reversed(range(n_modules)):
+        for stage_result in reversed(profile_results):
+            stage_no -= 1
+            module_result = stage_result.module_profile_results[i]
+            for invar in module_result.invar_names:
+                if invar not in eliminate_time:
+                    eliminate_time[invar] = stage_no
+            for required_idx in module_result.required_outvars_indices:
+                required_outvars.add(module_result.outvar_names[required_idx])
+
+    stage_no = 0
+    intermediate_size = None
+    for i in range(n_modules):
+        for stage_result in profile_results:
+            module_result = stage_result.module_profile_results[i]
+            for invar, size, donated in zip(module_result.invar_names,
+                                            module_result.invar_sizes,
+                                            module_result.donated_invars):
+                if invar not in env:
+                    env[invar] = size
+                else:
+                    assert env[invar] == size
+                if donated:
+                    del env[invar]
+            for outvar, size in zip(module_result.outvar_names,
+                                    module_result.outvar_sizes):
+                if outvar not in env:
+                    env[outvar] = size
+                else:
+                    assert env[outvar] == size
+            total_env_size = sum(env.values())
+            peak_memory = max(peak_memory,
+                              total_env_size + module_result.temp_buffer_size)
+            var_to_be_eliminated = []
+            for var in env:
+                if (var not in required_outvars and
+                    (var not in eliminate_time or
+                     eliminate_time[var] >= stage_no)):
+                    var_to_be_eliminated.append(var)
+            for var in var_to_be_eliminated:
+                del env[var]
+            stage_no += 1
+        intermediate_size = sum(env.values())
```

Comment:
why do we compute that in each iteration of the loop, but use it outside of the loop?

Reply:
Here I assume `n_modules == 2` and the `intermediate_size` captures the variables that are from the forward pass that is used in the backward pass.

Reply:
In case of that, doesn't it compute `intermediate_size` only when `i==0`?

Reply:
oh, you are right! thanks for catching this!